[chiralcentre] iP

chiralcentre opened the pull request:
### Felix
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Felix helps you to organise and keep track of your upcoming tasks. Felix is:

- text-based
- intuitive
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/chiralcentre/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is FREE!

Features:

- [ ] Managing upcoming tasks to be completed without a fixed deadline
- [ ] Managing upcoming tasks to be completed with a fixed deadline
- [ ] Managing upcoming events with a fixed start and end time
- [x] Managing reminders

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method in `Felix.java`:
```java
public static void main(String[] args) throws IOException, FelixException {
    Felix felix = new Felix();
    felix.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

chiralcentre committed:
Finished level-0

chiralcentre committed:
level-1

chiralcentre committed:
Added TaskList, Task classes

chiralcentre committed:
Implemented ability to mark and unmark tasks as done

chiralcentre committed:
Added support for tracking ToDos, Deadlines and Events

chiralcentre committed:
performed testing with I/O redirection

chiralcentre committed:
Added exception handling and enum for the different command types

chiralcentre committed:
Implemented deletion of tasks, and updated enums accordingly

chiralcentre committed:
Add support for saving tasks to hard disk, add support for generating tasks from lines in saved file

chiralcentre committed:
Merge branch 'branch-level-7'

chiralcentre committed:
Rename chatbot to Felix

chiralcentre committed:
change string formatting bugs in Event.java and Deadline.java

chiralcentre committed:
update logo in Felix.java

chiralcentre committed:
Implement parsing of strings into LocalDateTime objects for Event and Deadline tasks

chiralcentre committed:
Merge branch 'branch-level-8'

chiralcentre committed:
Extract more classes which inherit from an abstract Command class, and remove CommandType enum

chiralcentre committed:
Merge branch 'branch-A-MoreOOP'

chiralcentre committed:
Divide classes into packages

chiralcentre committed:
Merge branch 'branch-A-Packages'

chiralcentre committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

chiralcentre committed:
Add checkstyle plugin and enableAssetions = true to build.gradle

chiralcentre committed:
Merge branch 'branch-A-Gradle'

chiralcentre committed:
Add JUnit tests for Event and Parser methods

chiralcentre committed:
Merge branch 'branch-A-Junit'

chiralcentre committed:
Add JavaDoc comments for classes in task and utils packages

chiralcentre committed:
Merge branch 'branch-A-JavaDoc'

chiralcentre committed:
Refactor code in Parser and TaskList to conform to coding standard

chiralcentre committed:
Merge branch 'branch-A-CodingStandard'

chiralcentre committed:
Implement search by keyword functionality to find all tasks with specified keyword in description

chiralcentre committed:
Merge branch 'branch-Level-9'

yisiox commented on src/main/java/felix/utils/Parser.java:
```diff
@@ -0,0 +1,89 @@
+                String[] remainder = words[1].split(" /from | /to ");
+                task = new Event(remainder[0], remainder[1], remainder[2]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new FelixException("Command does not follow this format: event {description} /from {start_datetime} /to {end_datetime}");
```
Should this line be wrapped? I noticed the same issue in other places too.

chiralcentre replied:
Thank you for the feedback! Just to clarify, what do you mean by wrapping the line?

yisiox replied:
Hi, the course coding standard states that lines should not pass the soft limit of 110 lines, and a hard limit of 120 lines. The oracle coding standard states 80 lines. This line is really long so it could be changed to look (for e.g.) like
```java
throw new FelixException("Command does not follow this format: event {description} "
        + "/from {start_datetime} /to {end_datetime}");
```

yisiox commented on src/main/java/felix/utils/Storage.java:
```diff
@@ -0,0 +1,81 @@
+     * Generates a list of tasks from the storage file associated with the created Storage instance.
+     *
+     * @return list of tasks.
+     * @throws FelixException If file is not found, there is error reading from the file, or datetime is not in correct format.
```
Suggestion for formatting long comment:
```
@throws FelixException  If file is not found, there is error reading from the file, 
                        or datetime is not in correct format.
```

yisiox commented on src/main/java/felix/command/MarkCommand.java:
```diff
@@ -0,0 +1,26 @@
+import felix.task.Task;
+import felix.utils.TaskList;
+import felix.utils.Ui;
+import felix.utils.Storage;
```
Nit: perhaps add a line break here, and in some other places.

chiralcentre replied:
Thank you for the feedback! Just to clarify, do you mean adding a line break between every import statement?

yisiox commented on src/main/java/felix/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return task represented by file line.
+     */
+    public static Task createTaskFromFileLine(String line) {
+        String[] words = line.split(" \\| ");
+        Task task;
+        if (words[0].equals("T")) {
+            task = new ToDo(words[2]);
+        } else if (words[0].equals("D")) {
+            task = new Deadline(words[2], words[3]);
+        } else {
+            task = new Event(words[2], words[3], words[4]);
+        }
+        if (words[1].equals("X")) {
+            task.markAsDone();
+        }
+        return task;
+    }
```
I like how the `Task` class handles creation of the task objects so task types are hidden from `Storage`! Overlooked this myself. A nit may be a switch statement will make this look cleaner.

yisiox replied:
After some thought, I think this can be improved further by moving this to a separate `TaskFactory` class so as to not have `Task` "know" its subclasses.

chiralcentre replied:
Thank you for the feedback! Just to clarify, what is the purpose of the TaskFactory class?

yisiox replied:
Instead of having a `createTask` method inside the `Task` base class itself, which somewhat violates OOP since `Task` "knows" its subclasses, I made another class `TaskFactory` which is responsible for returning the `Task` of the relevant runtime type.

yisiox commented on src/main/java/felix/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Override
+    public String toString() {
+        return String.format("[%s] %s",this.getStatusIcon(),this.description);
```
Nit: spaces can be added after commas. I noticed the same issue in other places too.

chiralcentre replied:
Thank you for the feedback! I will fix this.

yisiox commented on src/main/java/felix/utils/Parser.java:
```diff
@@ -0,0 +1,89 @@
+
+    private Task generateTask(String[] words) throws FelixException {
+        Task task;
+        if (words[0].equals("todo")) {
+            try {
+                task = new ToDo(words[1]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new FelixException("The description of a todo cannot be empty");
+            }
+        } else if (words[0].equals("deadline")) {
+            try {
+                String[] remainder = words[1].split(" /by ");
+                task = new Deadline(remainder[0], remainder[1]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new FelixException("Command does not follow this format: deadline {description} by {end_datetime}");
+            } catch (DateTimeParseException e) {
+                throw new FelixException("datetime for deadline is not in the format \"yyyy-MM-dd HHmm\"");
+            }
+        } else {
+            try {
+                String[] remainder = words[1].split(" /from | /to ");
+                task = new Event(remainder[0], remainder[1], remainder[2]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new FelixException("Command does not follow this format: event {description} /from {start_datetime} /to {end_datetime}");
+            } catch (DateTimeParseException e) {
+                throw new FelixException("datetime not in the format \"yyyy-MM-dd HHmm\"");
+            }
+        }
```
Nit: is a switch statement cleaner here?

yisiox commented on src/main/java/felix/utils/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+        StringBuilder output = new StringBuilder();
+        for (int i = 0; i < this.tasks.size(); i++) {
+            output.append(this.tasks.get(i).toFileString());
+            if (i < this.tasks.size() - 1) output.append("\n");
```
Single statement conditionals should still be wrapped by curly brackets. I noticed the same issue in other places as well.

chiralcentre replied:
Thank you for the feedback! I will fix this.

yisiox commented on src/main/java/felix/utils/Ui.java:
```diff
@@ -0,0 +1,85 @@
+            + " |     \\\\  ___/|  |_|  |>    <  \n"
+            + " \\___  / \\___  >____/__/__/\\_ \\ \n"
+            + "     \\/      \\/              \\/ \n";
+    private static final String LINE_SEPARATOR = "__________________________"
```
Suggestion to make `LINE_SEPARATOR` size explicit: `LINE_SEPARATOR = "_".repeat(60);`

chiralcentre replied:
Thank you for the feedback! I will fix this.

yisiox reviewed (commented):
LGTM. Just some nits to fix.

chiralcentre committed:
Add JavaDoc comments and other fixes after running checkstyle

chiralcentre committed:
Merge branch 'A-CheckStyle'

chiralcentre committed:
Implement GUI using JavaFX

chiralcentre committed:
Merge branch 'branch-Level-10'

chiralcentre committed:
change mainClass in build.gradle

chiralcentre committed:
Add assertions to code
Assertion statements were added to UnmarkCommand.java, MarkCommand.java,
and AddTaskCommand.java to check postconditions.

chiralcentre committed:
Improve code quality
Explicitly specified length of line separator in Ui.java,
and added the message to be displayed when exiting the program
as a constant GOODBYE_MESSAGE to avoid magic literals.

chiralcentre committed:
Use Streams in TaskList.java

chiralcentre committed:
Merge pull request #1 from chiralcentre/branch-A-Assertions

Add assertions to code

chiralcentre committed:
Merge pull request #2 from chiralcentre/branch-A-CodeQuality

Improve code quality

chiralcentre committed:
Merge pull request #3 from chiralcentre/A-Streams

Use Streams in TaskList.java

alfaloo commented on src/main/java/felix/utils/Parser.java:
```diff
@@ -0,0 +1,101 @@
+     * @return Command.
+     * @throws FelixException If command is unrecognised or command formatting is wrong.
+     */
+    public Command getCommand(String line) throws FelixException {
```
Do you think its better to split this long method (more than 30 lines) into smaller more readable methods?

chiralcentre committed:
Implement update functionality for tasks in TaskList

chiralcentre committed:
Merge branch 'branch-C-Update'

chiralcentre committed:
Made small changes to code to beautify GUI

chiralcentre committed:
Fill in User Guide in /docs folder

chiralcentre committed:
Remove line breaks from docs/README.MD since they do not show up properly in Github Pages

chiralcentre committed:
Update root README

chiralcentre committed:
Update root README

chiralcentre committed:
Update root README

chiralcentre committed:
Update root README

chiralcentre committed:
Update root README

chiralcentre committed:
fix bug for todo
